### PR TITLE
Add executeTransactionCompletable method

### LINKED ISSUE #2126: async executeTransaction with Oberservables
It would be nice to have an async executeTransaction method in Realm-class which returns an Observable since we RxJava-users don't like callables ;)
Maybe something like:
```
Observable<void> executeTransaction(Realm.Transaction transaction)
```
Usage would be:
```
Observable<void> transaction = realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm bgRealm) {
                User user = bgRealm.createObject(User.class);
                user.setName("John");
                user.setEmail("john@corporation.com");
            }
        });
transaction.subscribe(new rx.Subscriber<T> {
           @Override public void onCompleted() {
                 //do your stuff here, if transaction was successful
            }

           @Override public void onError(Throwable e) {
                 //cleanup
            }

            @Override public void onNext(T t) {
                 //nothing.
            }
        });
```

### PR DESCRIPTION
**Pull request by jereksel**
fixes #2126

DynamicRealm's transactions are synchronous only, so I've only added Completable support for Realm.

### CODE REVIEW

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 1]
```diff
@@ -1586,6 +1588,18 @@ public void run() {
...
     }
 
+    /**
+     * Invokes transaction on worker thread and returns Completable
+     *
+     * @param transaction {@link io.realm.Realm.Transaction} to execute.
+     * @return {@link Completable} representing asynchronous transaction
+     * @throws UnsupportedOperationException if the required RxJava framework is not on the classpath.
+     */
+    public Completable executeTransactionCompletable(final Transaction transaction) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: I'm not 100% sure about this name. From the discussion in https://github.com/realm/realm-java/issues/2126#issuecomment-196844283 it seems that this should be renamed to `asCompletable(final Transaction transaction). This also seems to be more in line with the naming from e.g `RealmResults.asFlowable()`.

Or did you have some other reasoning for naming it this?

@nhachicha do you have any thoughts on this?

If we indeed end up wrapping an async transaction `asCompleteteAsync` would propably fit our current naming scheme better as well...Naming is hard :smile:

File: realm/realm-library/src/main/java/io/realm/rx/RealmObservableFactory.java [hunk 2]
```diff
@@ -592,6 +593,11 @@ public void run() {
         throw new RuntimeException("RealmQuery not supported yet.");
     }
 
+    @Override
+    public Completable from(Realm realm, Realm.Transaction transaction) {
+        return Completable.create(e -> realm.executeTransactionAsync(transaction, e::onComplete, e::onError));
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/rx/RealmObservableFactory.java]: I'm a bit torn whether we should use `executeTransaction` or `executeTransactionAsync` here. If we use `executeTransaction()` we force people to manually use `subscribeOn`, but somehow that feels more idiomatic. The problem with using Realms internal async system is that it e.g. becomes impossible to swap schedulers when testing.

Did you have any reason for choosing this over the synchronous variant?

In any case, if we keep this, we need to update the doc to say it throws if called from a non-looper thread.

@nhachicha / @Zhuinden Do you have any thoughts?

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 3]
```diff
@@ -1586,6 +1588,18 @@ public void run() {
         return new RealmAsyncTaskImpl(pendingTransaction, asyncTaskExecutor);
     }
 
+    /**
+     * Invokes transaction on worker thread and returns Completable
+     *
+     * @param transaction {@link io.realm.Realm.Transaction} to execute.
+     * @return {@link Completable} representing asynchronous transaction
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: representing the

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 4]
```diff
@@ -1586,6 +1588,18 @@ public void run() {
         return new RealmAsyncTaskImpl(pendingTransaction, asyncTaskExecutor);
     }
 
+    /**
+     * Invokes transaction on worker thread and returns Completable
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: `Invokes the transaction on a worker thread wrapping the result as a {@link Completable}.`

File: realm/realm-library/src/main/java/io/realm/rx/RealmObservableFactory.java (hunk 2 above)
Comment (Zhuinden) [reply realm/realm-library/src/main/java/io/realm/rx/RealmObservableFactory.java]: My thought is that this is a breaking change for anyone who implemented their custom `RxObservableFactory`.

But then again, it follows the https://github.com/realm/realm-java/issues/2126#issuecomment-196844283 proposition directly.

I think it should be possible to use `executeTransaction` vs `executeTransactionAsync` depending on whether the Realm is on a looper thread, which you can verify with `realm.isAutoRefresh()` or more-so the internal "capabilities" class should know whether this thread is able to use async transactions.

Comment (nhachicha) [reply realm/realm-library/src/main/java/io/realm/rx/RealmObservableFactory.java]: If you use `executeTransaction ` then you need to make sure the Realm you're using will be on the same scheduler's Thread (`subscribeOn`). I think `executeTransactionAsync` is appropriate here

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 1 above)
Comment (nhachicha) [reply realm/realm-library/src/main/java/io/realm/Realm.java]: `asCompletable` feels more aligned with current API

**Comment by jereksel** [comment]
@cmelchior I've fixed all comments

**Comment by idok595** [comment]
Hey There is no handling for the RealmAsyncTask! i think that when the Disoposable dispoded before it complete it is require to cancel the RealmAsyncTask have been created by the call "executeTransactionAsync"

**Pull request closed without merging**